pkg/autoscaler/metrics/stats_scraper.go

```diff
@@ -181,7 +189,113 @@ func (s *serviceScraper) Scrape(window time.Duration) (Stat, error) {
 	if readyPodsCount == 0 {
 		return emptyStat, nil
 	}
-	return s.scrapeService(window, readyPodsCount)
+
+	startTime := time.Now()
+	defer func() {
+		scrapeTime := time.Since(startTime)
+		pkgmetrics.RecordBatch(s.statsCtx, scrapeTimeM.M(float64(scrapeTime.Milliseconds())))
+	}()
+
+	if s.podsAddressable {
+		stat, err := s.scrapePods(readyPodsCount)
+		// Some pods were scraped, but not enough.
+		if err != errNoPodsScraped {
+			return stat, err
+		}
+		// Else fall back to service scrape.
+	}
+	stat, err := s.scrapeService(window, readyPodsCount)
+	if err == nil {
+		// If err == nil, this means that we failed to scrape all pods, but service worked
+		// thus it is probably a mesh case.
+		s.podsAddressable = false
+	}
+	return stat, err
+}
+
+func (s *serviceScraper) scrapePods(readyPods int) (Stat, error) {
+	pods, err := s.podAccessor.PodIPsByAge()
+	if err != nil {
+		return emptyStat, err
+	}
+	// Race condition when scaling to 0, where the check above
+	// for endpoint count worked, but here we had no ready pods.
+	if len(pods) == 0 {
+		return emptyStat, nil
+	}
+
+	frpc := float64(readyPods)
+	sampleSizeF := populationMeanSampleSize(frpc)
+	sampleSize := int(sampleSizeF)
+	results := make(chan Stat, sampleSize)
+
+	grp := errgroup.Group{}
+	idx := int32(-1)
+	for i := 0; i < sampleSize; i++ {
+		grp.Go(func() error {
+			for {
+				// Acquire next pod.
+				myIdx := int(atomic.AddInt32(&idx, 1))
```

Comment:
I don't get why we +1 for each run in the second for loop.

Reply:
We're just going down the list of pods and we need each pod to be scraped only once. Hence we need to select the next available.